src/text_shaper/directwrite_shaper.cpp

```diff
@@ -279,58 +284,137 @@ void directwrite_shaper::shape(font_key _font,
                                unicode::Script _script,
                                shape_result& _result)
 {
-    ComPtr<IDWriteTextAnalyzer> analyzer;
-    d->factory->CreateTextAnalyzer(analyzer.GetAddressOf());
-
-    // WCHAR const *textString = L""; // TODO
-    // UINT32 textLength; // TODO
-    // IDWriteFontFace *fontFace; // TODO: get from hashmap.at(key)
-    // BOOL isSideways = FALSE;
-    // BOOL isRightToLeft = FALSE;
-    // DWRITE_SCRIPT_ANALYSIS const* scriptAnalysis; // TODO: call to AnalyzeScript(...)
-    // WCHAR const* localeName; // TODO: current user locale
-    // IDWriteNumberSubstitution* numberSubstitution = NULL;
-
-    // auto EnableStdLigatures = DWRITE_FONT_FEATURE{ DWRITE_FONT_FEATURE_TAG_STANDARD_LIGATURES, TRUE };
-    // auto feature1 = DWRITE_TYPOGRAPHIC_FEATURES{
-    //     { &EnableStdLigatures },
-    //     1 // count
-    // };
-    // DWRITE_TYPOGRAPHIC_FEATURES* features[] = { &feature1 };
-    // UINT32 const featureRangeLengths[] = { textLength }; // The length of each feature range, in characters. The sum of all lengths should be equal to textLength.
-    // UINT32 featureRanges = 1; // the number of feature ranges
-
-    // auto maxGlyphCount = UINT32{256};
-    // auto clusterMap = new UINT16[maxGlyphCount];
-    // auto textProps = new DWRITE_SHAPING_TEXT_PROPERTIES[maxGlyphCount]; // needs allocation
-    // auto glyphIndices = new UINT16[maxGlyphCount];
-    // auto glyphProps = new DWRITE_SHAPING_GLYPH_PROPERTIES[maxGlyphCount];
-    // auto actualGlyphCount = UINT32{0};
-
-    // analyzer->GetGlyphs(textString,
-    //                     textLength,
-    //                     fontFace,
-    //                     isSideways,
-    //                     isRightToLeft,
-    //                     scriptAnalysis,
-    //                     localeName, numberSubstitution,
-    //                     &features/*FIXME: is this right?*/,
-    //                     featureRangeLengths,
-    //                     featureRanges,
-    //                     maxGlyphCount,
-    //                     clusterMap,
-    //                     textProps,
-    //                     glyphIndices,
-    //                     glyphProps,
-    //                     &actualGlyphCount);
-
+    std::wstring_convert<std::codecvt_utf8<char32_t>, char32_t> conv1;
```

Comment:
This is also dumb. I failed to convert `u32string` directly to `wstring`, so I had to do it like this. Hope there is a way to avoid such dumbness.

Reply:
Maybe the following should help here?

https://github.com/contour-terminal/libunicode/blob/master/src/unicode/convert.h#L332-L339

Mind, I have never done anything with UTF-16. So if the suggested way does not work, that this bug should be fixed (Out of my head, I do not know if I have unit-tested UTF-16, but if not, I should add some).

[libunicode](https://github.com/contour-terminal/libunicode/) is a Unicode library I specifically created to aid Contour's Unicode needs (especially the containing algorithms).

Reply:
If libunicode doesn’t handle “surrogate pair” correctly in UTF16, then the conversion won’t work. The better UTF-16 support is obviously not the focus of this PR so I think I will probably just leave it here, if you are OK with it. 

I do wonder why utf32 is used in this method. Does utf32 make things easier in harfbuzz or something? Just curious.